# DP: gm2 updates up to 20190820

git diff de2034bc023a89e45bec67e54d2a63982080c316 37ac5de83cefd414863a697866464ac9c8ac2140 | filterdiff --strip=2 --addoldprefix=a/src/ --addnewprefix=b/src/ --remove-timestamp | egrep -v '^(diff|index)'

--- a/src/gcc/gm2/ChangeLog
+++ b/src/gcc/gm2/ChangeLog
@@ -1,4 +1,59 @@
-2019-08-05      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+2019-08-20      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/Make-lang.in:  use -p parameter for mkdir.
+
+2019-08-20      Matthias Klose  <doko@cs.tu-berlin.de>
+
+	* gm2/Make-lang.in:  added dependancies for subdir
+          creation.
+
+2019-08-20      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/tools-src/buildsyntax:  removed function keyword
+          to be sh complient rather than bash.
+	* gm2/tools-src/makeSystem:  removed function keyword
+          to be sh complient rather than bash.
+	* gm2/tools-src/makeconfigure:  removed function keyword
+          to be sh complient rather than bash.
+	* gm2/tools-src/makeversion:  removed function keyword
+          to be sh complient rather than bash.  Also use = rather
+          than ==.
+
+2019-08-16      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/Make-lang.in:  (gm2/pge) added -lgcov.
+          (gm2/pg) added -lgcov.  (gm2/ppg)
+          added -lgcov.  (gm2/boot-bin/mc) added -lgcov.
+
+2019-08-13      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/Make-lang.in:  (check-gm2) removed, use inbuilt
+          version.  (lang_checks) include check-gm2.
+	* testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/examples/callingC/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/exceptions/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/extensions/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/imports/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/linking/libarchive/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/pim/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/gm2/types/run/pass/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+	* testsuite/lib/gm2.exp:
+          use rootme and rootme/gm2 rather than tmpdir.
+
+2019-08-13      Gaius Mulley <gaius.mulley@southwales.ac.uk>
+
+	* gm2/lang-specs.h:  added %d to the temporary
+          assembly file.
+
+2019-08-12      Gaius Mulley <gaius.mulley@southwales.ac.uk>
 
 	* New branch gcc_9_2_0_gm2
 	* gm2/configure.in:  changed release number to 1.9.2.
--- a/src/gcc/gm2/Make-lang.in
+++ b/src/gcc/gm2/Make-lang.in
@@ -312,11 +312,11 @@ $(objdir)/gm2/boot-bin:
 $(objdir)/gm2/ulm-lib-gm2:
 	mkdir $@
 
-$(objdir)/gm2/ulm-lib-gm2/sys:
-	mkdir $@
+$(objdir)/gm2/ulm-lib-gm2/sys: $(objdir)/gm2/ulm-lib-gm2
+	mkdir -p $@
 
-$(objdir)/gm2/ulm-lib-gm2/std:
-	mkdir $@
+$(objdir)/gm2/ulm-lib-gm2/std: $(objdir)/gm2/ulm-lib-gm2
+	mkdir -p $@
 
 $(objdir)/gm2/gm2-libs-pim:
 	mkdir $@
@@ -1757,7 +1757,9 @@ gm2/mc-boot-ch/$(SRC_PREFIX)%.o: gm2/mc-boot-ch/$(SRC_PREFIX)%.c
 mc-bootstrap: $(GM2_DIRS) mc-clean gm2/boot-bin/mc$(exeext)
 
 gm2/boot-bin/mc$(exeext): $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) $(LDFLAGS) gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-MC-BOOT-O) $(BUILD-MC-INTERFACE-O) $(LDFLAGS) \
+          gm2/mc-boot/main.o mcflex.o gm2/gm2-libs-boot/pthdummy.o -lm \
+          -lgcov
 
 gm2/mc-boot/main.o: $(M2LINK) $(srcdir)/gm2/init/mcinit
 	unset CC ; $(M2LINK) -s --gcc --exit --name mainmcinit.c $(srcdir)/gm2/init/mcinit
@@ -2104,7 +2106,7 @@ gm2/gm2-ppg-boot/$(SRC_PREFIX)%.o: $(srcdir)/gm2/gm2-compiler/%.mod $(MCDEPS) $(
               -I$(srcdir)/gm2/mc-boot-ch -g -c gm2/gm2-ppg-boot/$(SRC_PREFIX)$*.c -o $@
 
 gm2/ppg$(exeext): gm2/boot-bin/mc $(BUILD-PPG-O) $(BUILD-MC-INTERFACE-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PPG-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PPG-O) gm2/gm2-ppg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 
 gm2/gm2-ppg-boot/main.o: $(M2LINK) $(srcdir)/gm2/init/mcinit
 	unset CC ; $(M2LINK) -s --gcc --exit --name mainppginit.c $(srcdir)/gm2/init/ppginit
@@ -2150,7 +2152,7 @@ gm2/gm2-pg-boot/$(SRC_PREFIX)pg.o:  gm2/gm2-auto/pg.mod $(MCDEPS) $(BUILD-BOOT-H
 gm2/pg$(exeext): gm2/boot-bin/mc \
     $(BUILD-PG-O) $(GM2-PPG-MODS:%.mod=gm2/gm2-pg-boot/%.o) \
     $(BUILD-MC-INTERFACE-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PG-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PG-O) gm2/gm2-pg-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 
 gm2/gm2-auto/pginit:
 	sed -e 's/ppg/pg/' < $(srcdir)/gm2/init/ppginit > $@
@@ -2213,7 +2215,7 @@ gm2/gm2-pge-boot/$(SRC_PREFIX)pge.o:  gm2/gm2-auto/pge.mod $(MCDEPS) $(BUILD-BOO
 gm2/pge$(exeext): gm2/boot-bin/mc \
     $(BUILD-PGE-O) $(GM2-PPG-MODS:%.mod=gm2/gm2-pge-boot/%.o) \
     $(BUILD-MC-INTERFACE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o
-	$(CC) -g -o $@ $(BUILD-PGE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lm
+	$(CC) -g -o $@ $(BUILD-PGE-O) gm2/gm2-pge-boot/main.o gm2/gm2-libs-boot/pthdummy.o -lgcov -lm
 	$(srcdir)/gm2/tools-src/buildpg $(srcdir)/gm2/gm2-compiler/ppg.mod t > gm2/gm2-auto/t.bnf
 	./gm2/pge$(exeext) gm2/gm2-auto/t.bnf > gm2/gm2-auto/t1.mod
 	./gm2/pg$(exeext) gm2/gm2-auto/t.bnf > gm2/gm2-auto/t2.mod
@@ -2414,17 +2416,7 @@ check-modula-2: check-gm2
 check_modula-2: check-gm2
 check_modula-2: check-gm2
 
-GM2TESTSUITEDIR=testsuite
-
-check-gm2: $(GM2TESTSUITEDIR)/site.exp
-	-(rootme=`${PWD_COMMAND}`; export rootme; \
-	srcdir=`cd ${srcdir}; ${PWD_COMMAND}` ; export srcdir ; \
-	cd $(TESTSUITEDIR); \
-	EXPECT=${EXPECT} ; export EXPECT ; \
-	if [ -f $${rootme}/../expect/expect ] ; then  \
-	   TCL_LIBRARY=`cd .. ; cd ${srcdir}/../tcl/library ; ${PWD_COMMAND}` ; \
-	    export TCL_LIBRARY ; fi ; \
-	$(RUNTEST) --tool gm2 $(RUNTESTFLAGS))
+lang_checks += check-gm2
 
 check-gm2-local: $(GM2TESTSUITEDIR)/site.exp
 	-(rootme=`${PWD_COMMAND}`; export rootme; \
--- a/src/gcc/gm2/lang-specs.h
+++ b/src/gcc/gm2/lang-specs.h
@@ -31,7 +31,7 @@ Boston, MA 02110-1301, USA.  */
                        %{MT*} %{MF*} -quiet "
 
 #define GM2CC(INPUT,OUTPUT) \
-  "%{!fno-exceptions:cc1plus;:cc1} " GM2CC_OPTIONS " " INPUT " -o %b_m2.s \n\
+  "%{!fno-exceptions:cc1plus;:cc1} " GM2CC_OPTIONS " " INPUT " -o %d%b_m2.s \n\
   " AS("%b_m2.s",OUTPUT) " "
 
 #define GM2LCC(OBJECT,LST) \
--- a/src/gcc/gm2/tools-src/buildsyntax
+++ b/src/gcc/gm2/tools-src/buildsyntax
@@ -8,21 +8,21 @@
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation; either version 3, or (at your option)
 # any later version.
-# 
+#
 # GNU Modula-2 is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-# 
+#
 # You should have received a copy of the GNU General Public License
 # along with GNU Modula-2; see the file COPYING.  If not, write to the
 # Free Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA
-# 02110-1301, USA. 
+# 02110-1301, USA.
 #
 
 TMPFILE=/tmp/t.$$
 
-function buildpass1 () {
+buildpass1 () {
     if [ -x ../bin3/pg ] ; then
         if ../bin3/pg ../bnf/m2.bnf > $TMPFILE ; then
            cat $TMPFILE | sed -e "s/WHICHPASS1/TRUE/" > P1SyntaxCheck.mod
@@ -36,7 +36,7 @@ function buildpass1 () {
     fi
 }
 
-function buildpass2 () {
+buildpass2 () {
     if [ -x ../bin3/pg ] ; then
         if ../bin3/pg ../bnf/m2-2.bnf > $TMPFILE ; then
            cat $TMPFILE | sed -e "s/WHICHPASS2/TRUE/" > P2Build.mod
@@ -50,7 +50,7 @@ function buildpass2 () {
     fi
 }
 
-function buildpass3 () {
+buildpass3 () {
     if [ -x ../bin3/pg ] ; then
         if ../bin3/pg ../bnf/m2-3.bnf > $TMPFILE ; then
            cat $TMPFILE | sed -e "s/WHICHPASS3/TRUE/" > P3Build.mod
@@ -64,7 +64,7 @@ function buildpass3 () {
     fi
 }
 
-function buildpassH () {
+buildpassH () {
     if [ -x ../bin3/pg ] ; then
         if ../bin3/pg ../bnf/m2-h.bnf > $TMPFILE ; then
            cat $TMPFILE | sed -e "s/WHICHPASS3/TRUE/" > PHBuild.mod
@@ -82,7 +82,7 @@ function buildpassH () {
 #  buildsyntax - build the pass 1 parser with error recovery
 #
 
-function buildsyntax () {
+buildsyntax () {
    buildpass1
    buildpass2
    buildpass3
--- a/src/gcc/gm2/tools-src/makeSystem
+++ b/src/gcc/gm2/tools-src/makeSystem
@@ -19,7 +19,7 @@
 # Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. *)
 
 
-function Usage () {
+Usage () {
    echo "Usage: makesystem dialectflag SYSTEM.def SYSTEM.mod librarypath compiler"
 }
 
@@ -41,7 +41,7 @@ if [ "$DIALECT" != "-fiso" -a "$DIALECT" != "-fpim" ] ; then
    exit 1
 fi
 
-function displayExportedTypes () {
+displayExportedTypes () {
    n=1
    c=0
    for i in ${types} ; do
@@ -60,24 +60,24 @@ function displayExportedTypes () {
    echo " " >> ${OUTPUTFILE}
 }
 
-function displayBuiltinTypes () {
+displayBuiltinTypes () {
    for i in ${types} ; do
       echo "   $i ; " >> ${OUTPUTFILE}
    done
 }
 
-function displayStart () {
+displayStart () {
    sed -e "1,/@SYSTEM_DATATYPES@/!d" < ${SYSTEMDEF} | \
    sed -e "/@SYSTEM_DATATYPES@/d" >> ${OUTPUTFILE}
 }
 
-function displayMiddle () {
+displayMiddle () {
    sed -e "1,/@SYSTEM_DATATYPES@/d" < ${SYSTEMDEF} | \
    sed -e "1,/@SYSTEM_TYPES@/!d" | \
    sed -e "/@SYSTEM_TYPES@/d" >> ${OUTPUTFILE}
 }
 
-function displayEnd () {
+displayEnd () {
    sed -e "1,/@SYSTEM_TYPES@/d" < ${SYSTEMDEF} >> ${OUTPUTFILE}
 }
 
--- a/src/gcc/gm2/tools-src/makeconfigure
+++ b/src/gcc/gm2/tools-src/makeconfigure
@@ -8,16 +8,16 @@
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation; either version 3, or (at your option)
 # any later version.
-# 
+#
 # GNU Modula-2 is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-# 
+#
 # You should have received a copy of the GNU General Public License
 # along with GNU Modula-2; see the file COPYING.  If not, write to the
 # Free Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA
-# 02110-1301, USA. 
+# 02110-1301, USA.
 #
 
 #
@@ -25,13 +25,13 @@
 # file:
 #
 # gm2s/386-gas/M2Configure.def
-# 
+#
 
 #
 # create configuration constant
 #
 
-function EmitConst () {
+EmitConst () {
    echo "  " $1 "=" $2 ";  (* automatically generated *)" >> $DESTINATION
    echo "" >> $DESTINATION
 }
@@ -40,7 +40,7 @@ function EmitConst () {
 # remove the old configuration file and create new empty ones
 #
 
-function Initialize () {
+Initialize () {
    /bin/rm -f $DESTINATION
    touch $DESTINATION
 }
@@ -50,7 +50,7 @@ function Initialize () {
 # ModulePrologue - write the start of the definition module.
 #
 
-function ModulePrologue () {
+ModulePrologue () {
    cat >> $DESTINATION << EOF
 DEFINITION MODULE M2Configure ;
 
@@ -113,7 +113,7 @@ EOF
 # ModuleEpilogue - terminates the definition module.
 #
 
-function ModuleEpilogue () {
+ModuleEpilogue () {
    echo "" >> $DESTINATION
    echo "END M2Configure." >> $DESTINATION
 }
@@ -123,7 +123,7 @@ function ModuleEpilogue () {
 # work out whether we are compiling on a xenix system
 #
 
-function IsSystemXenix () {
+IsSystemXenix () {
    echo -n "are we compiling under xenix..."
    cat >> $DESTINATION << EOF3
 (*
@@ -142,7 +142,7 @@ EOF3
 }
 
 
-function UseUnderscoreForC () {
+UseUnderscoreForC () {
    cat >> $DESTINATION << EOF
 (*
    UseUnderscoreForC - if true then the C compiler uses _ in front of
@@ -168,7 +168,7 @@ EOF
 }
 
 
-function UseDotForGDBLabels () {
+UseDotForGDBLabels () {
    cat >> $DESTINATION << EOF
 (*
    UseDotForGDBLabels - if true then the C compiler produced .LBB2: for the
@@ -195,7 +195,7 @@ EOF
 }
 
 
-function UseDotForJumpLabels () {
+UseDotForJumpLabels () {
    cat >> $DESTINATION << EOF
 (*
    UseDotForJumpLabels - if true then the C compiler produced .L2: for jump
@@ -222,12 +222,12 @@ EOF
 }
 
 
-function secondword () {
+secondword () {
    echo $2
 }
 
 
-function AlignmentSize () {
+AlignmentSize () {
    cat >> $DESTINATION << EOF
 (*
    AlignmentSize - returns the default alignment size used.
@@ -247,7 +247,7 @@ EOF
 }
 
 
-function ActivationRecordOffset () {
+ActivationRecordOffset () {
    cat >> $DESTINATION << EOF
 (*
    ActivationRecordOffset - the number of words relative to frame pointer
@@ -260,7 +260,7 @@ EOF
 }
 
 
-function UseShortStabLineNumbers () {
+UseShortStabLineNumbers () {
    cat >> $DESTINATION << EOF
 (*
    UseShortStabLineNumbers - if true then the C compiler produced .stabd 68,0,2
@@ -288,7 +288,7 @@ EOF
 }
 
 
-function DefaultLibraryPath () {
+DefaultLibraryPath () {
    cat >> $DESTINATION << EOF
 (*
    DefaultLibraryPath - defermines the default library path and creates a
@@ -302,7 +302,7 @@ EOF
 }
 
 
-function UsingGCCBackend () {
+UsingGCCBackend () {
    cat >> $DESTINATION << EOF
 (*
    UsingGCCBackend - is the compiler being built with the GCC code generator?
--- a/src/gcc/gm2/tools-src/makeversion
+++ b/src/gcc/gm2/tools-src/makeversion
@@ -29,7 +29,7 @@
 
 
 
-function doM2 () {
+doM2 () {
 /bin/rm -f M2Version.mod
 
 cat << EOF >> M2Version.mod
@@ -71,7 +71,7 @@ EOF
 }
 
 
-function doC () {
+doC () {
 
 cat <<EOF > gm2version.c
 /* Generated by makeversion - do not edit */
@@ -95,7 +95,7 @@ EOF
 }
 
 
-function doCP () {
+doCP () {
 
 cat <<EOF > gm2version.c
 /* Generated by makeversion - do not edit */
@@ -119,7 +119,7 @@ EOF
 }
 
 
-function doTexi () {
+doTexi () {
 
 cat <<EOF > version.texi
 
@@ -135,7 +135,7 @@ EOF
 
 progname=$0
 
-function usage () {
+usage () {
    cat <<EOF
 Usage: $progname [-m][-c][-t]
    -m  generates a Modula-2 module M2Version.mod
@@ -206,7 +206,7 @@ fi
 #  $(srcdir)/gm2/gm2-libs/config-host.in
 #
 
-if [ "`grep AC_INIT ${SRCDIR}/gm2/gm2-libs/config-host.in | grep ${GM2VERSION}`" == "" ] ; then
+if [ "`grep AC_INIT ${SRCDIR}/gm2/gm2-libs/config-host.in | grep ${GM2VERSION}`" = "" ] ; then
    echo "makeversion has found a consistency error:  the ${SRCDIR}/gm2/gm2-libs/config-host.in does not match the GNU Modula-2 release number"
    exit 1
 fi
--- a/src/gcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/calling-c/datatypes/unbounded/run/pass/gm2.exp
@@ -27,12 +27,12 @@ set gm2src ${srcdir}/../gm2
 
 gm2_init_pim "${srcdir}/gm2/calling-c/datatypes/unbounded/run/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with "c.o"
 
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/calling-c/datatypes/unbounded/run/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/calling-c/datatypes/unbounded/run/pass/c.c -o c.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
--- a/src/gcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/examples/callingC/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -26,11 +26,11 @@ set gm2src ${srcdir}/../gm2
 
 gm2_init_iso "$srcdir/gm2/examples/callingC/run/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with c.o
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/examples/callingC/run/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/examples/callingC/run/pass/c.c -o c.o]
 
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
--- a/src/gcc/testsuite/gm2/exceptions/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/exceptions/run/pass/gm2.exp
@@ -25,11 +25,11 @@ load_lib prune.exp
 load_lib gm2-torture.exp
 
 
-set XGPP [lookfor_file $tmpdir xg++];
+set XGPP [lookfor_file $rootme xg++];
 
 set output [exec rm -f cpp.o mycpp.o]
-set output [exec ${XGPP} -B[file dirname $tmpdir] -g -c $srcdir/gm2/exceptions/run/pass/cpp.cpp]
-set output [exec ${XGPP} -B[file dirname $tmpdir] -g -c $srcdir/gm2/exceptions/run/pass/mycpp.cpp]
+set output [exec ${XGPP} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/exceptions/run/pass/cpp.cpp]
+set output [exec ${XGPP} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/exceptions/run/pass/mycpp.cpp]
 
 #
 #  notice this uses PIM libraries with exceptions - this is a useful test.
--- a/src/gcc/testsuite/gm2/extensions/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/extensions/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -28,10 +28,10 @@ gm2_init_pim "${srcdir}/gm2/extensions/run/pass" -fsoft-check-all
 
 gm2_link_with cvararg.o
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 set output [exec rm -f cvararg.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/extensions/run/pass/cvararg.c -o cvararg.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/extensions/run/pass/cvararg.c -o cvararg.o]
 
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
--- a/src/gcc/testsuite/gm2/imports/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/imports/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -29,7 +29,7 @@ gm2_init_pim "${srcdir}/gm2/imports/run/pass"
 gm2_link_with "-lgm2"
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
-    set output [exec ../xgm2 -B../ -g -c "-I${gccpath}/libgm2/libpim:${gm2src}/gm2-libs:${srcdir}/gm2/imports/run/pass" -fpim ${srcdir}/gm2/imports/run/pass/c.mod]
+    set output [exec ../../xgm2 -B../../ -g -c "-I${gccpath}/libgm2/libpim:${gm2src}/gm2-libs:${srcdir}/gm2/imports/run/pass" -fpim ${srcdir}/gm2/imports/run/pass/c.mod]
 
     # If we're only testing specific files and this isn't one of them, skip it.
     if ![runtest_file_p $runtests $testcase] then {
--- a/src/gcc/testsuite/gm2/linking/libarchive/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/linking/libarchive/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -24,13 +24,13 @@ load_lib gm2-torture.exp
 
 set gm2src ${srcdir}/../gm2
 
-gm2_init_iso "${srcdir}/gm2/linking/libarchive/pass" 
+gm2_init_iso "${srcdir}/gm2/linking/libarchive/pass"
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 gm2_link_with c.o
 set output [exec rm -f c.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/linking/libarchive/pass/c.c -o c.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/linking/libarchive/pass/c.c -o c.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
--- a/src/gcc/testsuite/gm2/pim/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/pim/run/pass/gm2.exp
@@ -27,7 +27,7 @@ set gm2src ${srcdir}/../gm2
 gm2_init_pim "${srcdir}/gm2/pim/run/pass" -fsoft-check-all
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
-    set output [exec ../xgm2 -B../ -g -c -I$srcdir/../gm2/gm2-libs -I$srcdir/gm2/pim/run/pass -I$srcdir/../gm2/gm2-compiler -I../gm2/gm2-libs -I../gm2/gm2-compiler -fpim $srcdir/gm2/pim/run/pass/sys.mod]
+    set output [exec ../../xgm2 -B../../ -g -c -I$srcdir/../gm2/gm2-libs -I$srcdir/gm2/pim/run/pass -I$srcdir/../gm2/gm2-compiler -I../gm2/gm2-libs -I../gm2/gm2-compiler -fpim $srcdir/gm2/pim/run/pass/sys.mod]
 
     # If we're only testing specific files and this isn't one of them, skip it.
     if ![runtest_file_p $runtests $testcase] then {
--- a/src/gcc/testsuite/gm2/types/run/pass/gm2.exp
+++ b/src/gcc/testsuite/gm2/types/run/pass/gm2.exp
@@ -13,7 +13,7 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA. 
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
 if $tracelevel then {
     strace $tracelevel
@@ -24,14 +24,14 @@ load_lib gm2-torture.exp
 
 set gm2src ${srcdir}/../gm2
 
-gm2_init_pim "${srcdir}/gm2/types/run/pass" 
+gm2_init_pim "${srcdir}/gm2/types/run/pass"
 
 gm2_link_with d.o
 
-set XGCC [lookfor_file $tmpdir xgcc];
+set XGCC [lookfor_file $rootme xgcc];
 
 set output [exec rm -f d.o]
-set output [exec ${XGCC} -B[file dirname $tmpdir] -g -c $srcdir/gm2/types/run/pass/d.c -o d.o]
+set output [exec ${XGCC} -B[file dirname $rootme]/gcc -g -c $srcdir/gm2/types/run/pass/d.c -o d.o]
 
 foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     # If we're only testing specific files and this isn't one of them, skip it.
@@ -42,4 +42,4 @@ foreach testcase [lsort [glob -nocomplain $srcdir/$subdir/*.mod]] {
     gm2-torture-execute $testcase "" "pass"
 }
 
-set output [exec rm -f d.o]
\ No newline at end of file
+set output [exec rm -f d.o]
--- a/src/gcc/testsuite/lib/gm2.exp
+++ b/src/gcc/testsuite/lib/gm2.exp
@@ -92,8 +92,8 @@ proc gm2_init { args } {
     if { $gm2_initialized == 1 } { return; }
 
     set gm2_link_libraries "";
-    set GCC_UNDER_TEST [lookfor_file $tmpdir xgm2];
-    append GCC_UNDER_TEST " " -B[file dirname $tmpdir] " " ${args};
+    set GCC_UNDER_TEST [lookfor_file $rootme xgm2];
+    append GCC_UNDER_TEST " " -B[file dirname $rootme]/gcc " " ${args};
     append GCC_UNDER_TEST " " -fno-diagnostics-show-caret
     append GCC_UNDER_TEST " " -fno-diagnostics-show-line-numbers
     append GCC_UNDER_TEST " " -fdiagnostics-color=never
@@ -123,7 +123,6 @@ proc gm2_init { args } {
 #
 
 proc gm2_target_compile_default { source dest type options } {
-    global tmpdir;
     global gluefile wrap_flags;
     global GCC_UNDER_TEST;
     global TOOL_OPTIONS;
